runtime/stdlib/account.go

```diff
@@ -791,6 +800,245 @@ func newPublicAccountContractsValue(
+			}
+
+			recipientValue := invocation.Arguments[2].(interpreter.AddressValue)
+			if !ok {
+				panic(errors.NewUnreachableError())
+			}
+
+			inter := invocation.Interpreter
+			getLocationRange := invocation.GetLocationRange
+
+			handler.EmitEvent(
+				inter,
+				AccountInboxPublishEventType,
+				[]interpreter.Value{
+					providerValue,
+					recipientValue,
+					nameValue,
+					interpreter.NewTypeValue(gauge, publishedValue.StaticType(inter)),
+				},
+				getLocationRange,
+			)
+
+			publishedValue = publishedValue.Transfer(
+				inter,
+				getLocationRange,
+				atree.Address(address),
+				true,
+				nil,
+			)
+
+			recipient := recipientValue.Transfer(
+				inter,
+				getLocationRange,
+				atree.Address(address),
+				true,
+				nil,
+			)
+
+			// we need to store both a value and an intended recipient for each name,
+			// so we do two writes to represent each published value.
+			inter.WriteStored(address, inboxStorageDomain, valuePath(nameValue.Str), publishedValue)
+			inter.WriteStored(address, inboxStorageDomain, recipientPath(nameValue.Str), recipient)
+
+			return interpreter.NewVoidValue(gauge)
+		},
+		sema.AuthAccountTypeInboxPublishFunctionType,
+	)
+}
+
+func accountInboxUnpublishFunction(
+	gauge common.MemoryGauge,
+	handler EventEmitter,
+	address common.Address,
+	providerValue interpreter.AddressValue,
+) *interpreter.HostFunctionValue {
+	return interpreter.NewHostFunctionValue(
+		gauge,
+		func(invocation interpreter.Invocation) interpreter.Value {
+			nameValue, ok := invocation.Arguments[0].(*interpreter.StringValue)
+			if !ok {
+				panic(errors.NewUnreachableError())
+			}
+
+			inter := invocation.Interpreter
+			getLocationRange := invocation.GetLocationRange
+
+			publishedValue := inter.ReadStored(address, inboxStorageDomain, valuePath(nameValue.Str))
+
+			if publishedValue == nil {
+				return interpreter.NewNilValue(gauge)
+			}
+
+			typeParameterPair := invocation.TypeParameterTypes.Oldest()
+			if typeParameterPair == nil {
+				panic(errors.NewUnreachableError())
+			}
+
+			ty := sema.NewCapabilityType(gauge, typeParameterPair.Value)
+			publishedType := publishedValue.StaticType(invocation.Interpreter)
+			if !inter.IsSubTypeOfSemaType(publishedType, ty) {
+				panic(interpreter.ForceCastTypeMismatchError{
+					ExpectedType:  ty,
+					ActualType:    inter.MustConvertStaticToSemaType(publishedType),
+					LocationRange: getLocationRange(),
+				})
+			}
+
+			handler.EmitEvent(
+				inter,
+				AccountInboxRemoveEventType,
+				[]interpreter.Value{
+					providerValue,
+					providerValue,
+					nameValue,
+				},
+				getLocationRange,
+			)
+
+			inter.WriteStored(address, inboxStorageDomain, valuePath(nameValue.Str), nil)
+			inter.WriteStored(address, inboxStorageDomain, recipientPath(nameValue.Str), nil)
+
+			publishedValue = publishedValue.Transfer(
+				inter,
+				getLocationRange,
+				atree.Address(address),
+				true,
+				nil,
+			)
```

Comment:
This order of the operations, and the transfer isn't quite right. You can follow how `load` is implemented, i.e. what `authAccountReadFunction` does, see the comment there: first transfer to the stack (empty address), then write nil to storage.